executor/merge_join.go

```diff
@@ -249,72 +221,69 @@ func (e *MergeJoinExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
 			return errors.Trace(err)
 		}
 	}
-	for {
-		if chk.NumRows() == e.maxChunkSize {
-			return nil
-		}
 
-		// reach here means there is no more data in "e.resultChunk"
-		hasMore, err := e.joinToChunk(chk)
+	for chk.NumRows() < e.maxChunkSize {
+		hasMore, err := e.joinToChunk(ctx, chk)
 		if err != nil || !hasMore {
 			return errors.Trace(err)
 		}
 	}
+	return nil
 }
 
-func (e *MergeJoinExec) joinToChunk(chk *chunk.Chunk) (hasMore bool, err error) {
-	for e.outerRows != nil {
-		// Here we set default cmpResult to -1 to emit unmatched outer rows.
+func (e *MergeJoinExec) joinToChunk(ctx context.Context, chk *chunk.Chunk) (hasMore bool, err error) {
+	for {
+		if e.outerTable.row == e.outerTable.iter.End() {
+			err = e.fetchNextOuterRows(ctx)
+			if err != nil || e.outerTable.chk.NumRows() == 0 {
+				return false, errors.Trace(err)
+			}
+		}
+
 		cmpResult := -1
-		if e.innerRows != nil {
-			cmpResult = compareChunkRow(e.compareFuncs, e.outerRows[0], e.innerRows[0], e.outerKeys, e.innerKeys)
+		if e.outerTable.selected[e.outerTable.row.Idx()] && len(e.innerRows) > 0 {
+			cmpResult = compareChunkRow(e.compareFuncs, e.outerTable.row, e.innerRows[0], e.outerTable.keys, e.innerTable.joinKeys)
 		}
+
 		if cmpResult > 0 {
 			if err = e.fetchNextInnerRows(); err != nil {
 				return false, errors.Trace(err)
 			}
 			continue
 		}
+
 		if cmpResult < 0 {
-			for e.outerIter4Row.Current() != e.outerIter4Row.End() {
-				if err = e.resultGenerator.emitToChunk(e.outerIter4Row.Current(), nil, chk); err != nil {
-					return false, errors.Trace(err)
-				}
-				e.outerIter4Row.Next()
-				if chk.NumRows() == e.maxChunkSize {
-					return true, errors.Trace(err)
-				}
-			}
-			if err = e.fetchNextOuterRows(); err != nil {
-				return false, errors.Trace(err)
-			}
-			continue
-		}
-		for e.outerIter4Row.Current() != e.outerIter4Row.End() {
-			if err = e.resultGenerator.emitToChunk(e.outerIter4Row.Current(), e.innerIter4Row, chk); err != nil {
+			err = e.resultGenerator.emitToChunk(e.outerTable.row, nil, chk)
+			if err != nil {
 				return false, errors.Trace(err)
 			}
-			if e.innerIter4Row.Current() == e.innerIter4Row.End() {
-				e.outerIter4Row.Next()
-				e.innerIter4Row.Begin()
-			}
-			if chk.NumRows() == e.maxChunkSize {
-				return true, errors.Trace(err)
+
+			e.outerTable.row = e.outerTable.iter.Next()
+
+			if chk.NumRows() >= e.maxChunkSize {
```

Comment:
s/ >=/ ==
since chk.NumRows can be promised not greater than maxChunkSize.

Reply:
I think `>=` is more safe.